ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml

```diff
@@ -0,0 +1,255 @@
+
+    property string inputTokenKey
+    onInputTokenKeyChanged: Qt.callLater(d.setSelectedHoldingId, inputTokenKey, Constants.TokenType.ERC20)
+    property string inputTokenAmount
+    onInputTokenAmountChanged: Qt.callLater(() => amountToSendInput.input.text =
+                                            Number(inputTokenAmount).toLocaleString(Qt.locale(), 'f', -128))
+
+    property int swapSide: SwapInputPanel.SwapSide.Pay
+
+    // output API
+    readonly property string selectedHoldingId: d.selectedHoldingId
+    readonly property double cryptoValue: amountToSendInput.cryptoValueToSendFloat
+    readonly property string cryptoValueRaw: amountToSendInput.cryptoValueToSend
+    readonly property bool cryptoValueValid: amountToSendInput.inputNumberValid
+
+    // visual properties
+    property int swapExchangeButtonWidth: 44
+    property string caption: swapSide === SwapInputPanel.SwapSide.Pay ? qsTr("Pay") : qsTr("Receive")
+
+    enum SwapSide {
+        Pay = 0,
+        Receive = 1
+    }
+
+    padding: Style.current.padding
+
+    // by design
+    implicitWidth: 492
+    implicitHeight: 131
+
+    QtObject {
+        id: d
+
+        function setSelectedHoldingId(holdingId, holdingType) {
+            let holding = root.store.getHolding(holdingId, holdingType)
+            d.selectedHoldingId = holdingId
+            d.setSelectedHolding(holding, holdingType)
+        }
+
+        function setSelectedHolding(holding, holdingType) {
+            d.selectedHoldingType = holdingType
+            d.selectedHolding = holding
+            let selectorHolding = root.store.holdingToSelectorHolding(holding, holdingType)
+            holdingSelector.setSelectedItem(selectorHolding, holdingType)
+        }
+
+        property var selectedHolding: null
+        property var selectedHoldingType: Constants.TokenType.Unknown
+        property string selectedHoldingId
+
+        readonly property bool isSelectedHoldingValidAsset: !!selectedHolding && selectedHoldingType === Constants.TokenType.ERC20
+        readonly property double maxFiatBalance: isSelectedHoldingValidAsset ? selectedHolding.currentCurrencyBalance : 0
+        readonly property double maxCryptoBalance: isSelectedHoldingValidAsset ? selectedHolding.currentBalance : 0
+        readonly property double maxInputBalance: amountToSendInput.inputIsFiat ? maxFiatBalance : maxCryptoBalance
+        readonly property string inputSymbol: amountToSendInput.inputIsFiat ? root.store.currencyStore.currentCurrency :
+                                                                              !!d.selectedHolding && !!d.selectedHolding.symbol ? d.selectedHolding.symbol: ""
+
+        function prepareForMaxSend(value, symbol) {
+            if (symbol !== Constants.ethToken) {
+                return value
+            }
+
+            return value - Math.max(0.0001, Math.min(0.01, value * 0.1))
```

Comment:
can we add a comment as to why we use this mathematical formula here?

Reply:
Yup, if only I knew why :laughing: this is copy&pasted from SendModal

Reply:
lol I dont know either. @dlipicar do you?

Reply:
https://github.com/status-im/status-desktop/issues/12215#issuecomment-1729465639

Reply:
what about receiving the max fees as ETH value in this component? I think the swapModal could be more accurate because the fees are calculated in the same step, so it could be possible to consider the fees in the max value calculations

Reply:
Ah but that logic is applied to the "send" transactions, I wonder if we should apply the same here (we are not sending anything, merely exchanging one for another)

Reply:
@alaibe any idea if that same logic should apply here for the Swap as well?

Reply:
This logic, is here to make sure there is enough eth to pay for the gas.
Context, when making a transaction, whatever the type: swap/bridge/send, you need eth to pay for the gas.

So if you plan to swap DAI for USDT, this logic shouldn't be applied, but you still need a bit of eth to create a transaction.

If you plan to swap ETH for USDT, or send ETH or bridge ETH, you cannot send your full balance as you need a bit of eth to pay for the gas, this is where this function is used, to prevent send your full balance and keep a buffer for the gas.

At this stage we don't have any estimation done yet, so this is why this should be more than enough, but at same time, very little values

Reply:
OK, we're keeping it :) and I made it so that this rounding function is also reflected in the validator's max value

Reply:
lets add a comment here to explain the reasoning for this formula in code please!